File: advent-of-code/2023/day-9/main.py
import sys

data = open(sys.argv[1]).read().splitlines()
input = [list(map(int, item.strip().split())) for item in data]


def diff(v: list[int]):
    res = []
    for i in range(1, len(v)):
        res.append(v[i] - v[i - 1])

    return res


def extrapolate2part(nums: list[int]):
    diff = 0
    startNumber = nums[-1]

    for i in range(len(nums) - 1, 0, -1):
        diff = nums[i - 1] - startNumber
        startNumber = diff

    return diff


cont = 0
part2 = 0
for v in input:
    startValues = []
    startValues.append(v[0])

    cont += v[-1]
    res = diff(v)
    cont += res[-1]
    startValues.append(res[0])
    while all([e == res[0] for e in res]) is False:
        res = diff(res)
        cont += res[-1]
        startValues.append(res[0])

    part2 += extrapolate2part(startValues)


print(cont)
print(part2)

File: advent-of-code/2023/day-9/main.cpp
#include <bits/stdc++.h>

using namespace std;

using vi = vector<int>;

vi convertToNumbers(const string &line) {
    vi ans;
    int number;
    stringstream ss(line);

    while(ss>>number)
        ans.push_back(number);

    return ans;
}

vector<vi> loadData(const string &name) {
    fstream input{name};
    string line;
    vector<vi> data;

    while(getline(input, line)) {
        data.push_back(convertToNumbers(line));
    }

    return data;
}

vi diff(const vi &num) {
    vi temp;
    temp.reserve(num.size() - 1);

    for(int i = 1; i < num.size(); ++i) {
        temp.push_back(num[i] - num[i - 1]);
    }

    return temp;
}

int extrapolate2part(vi num) {
    int diff = 0;

    int startNumber = num.back();

    for(int i = num.size() - 2; i >= 0; i-=1) {
        diff = num[i] - startNumber;
        startNumber = diff;
    }


    return diff;
}

int main (int argc, char *argv[]) {
    if(argc > 2 || argc == 1)
        cerr<<"missing argument"<<endl;

    int part1 = 0;
    int part2 = 0;
    for(auto v: loadData(string(argv[1]))) {
        vi startValue;
        startValue.push_back(v.front());

        part1 += v.back();
        vi res = diff(v);
        part1 += res.back();
        startValue.push_back(res.front());

        while(all_of(res.begin(), res.end(), [res](int x) { return x == res.back(); }) == false) {
            res = diff(res);
            part1 += res.back();
            startValue.push_back(res.front());
        }

        part2 += extrapolate2part(startValue);
    }

    cout<<part1<<endl;
    cout<<part2;

    return 0;
}